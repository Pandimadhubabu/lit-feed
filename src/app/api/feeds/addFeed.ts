import * as feedModels from "@/app/api/models/feeds";
import { Feed } from "@/types";

export async function addFeed({ body }: { body: Feed }) {
  const newFeed = await feedModels.addFeed(body);

  return {
    data: newFeed,
    status: 201,
    message: "Feed added",
  };
}
