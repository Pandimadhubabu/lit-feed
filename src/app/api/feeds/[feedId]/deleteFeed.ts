import * as feedModels from "@/models/feeds";
import { Feed } from "@/types";

export async function deleteFeed({
  params: { feedId },
}: {
  params: { feedId: Feed["id"] };
}) {
  const data = await feedModels.deleteFeed(feedId);

  return {
    status: 200,
    data,
    message: "Feed deleted",
  };
}
