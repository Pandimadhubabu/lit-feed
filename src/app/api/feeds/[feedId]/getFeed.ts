import * as feedModels from "@/models/feeds";
import { Feed } from "@/types";

export async function getFeed({
  params: { feedId },
}: {
  params: { feedId: Feed["id"] };
}) {
  const data = await feedModels.getFeed(feedId);

  return {
    status: 200,
    data,
    message: "Feed retrieved",
  };
}
