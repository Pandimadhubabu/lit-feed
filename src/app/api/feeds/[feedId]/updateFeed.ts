import * as feedModels from "@/models/feeds";
import { Feed } from "@/types";

export async function updateFeed({
  body,
  params: { feedId },
}: {
  body: Feed;
  params: { feedId: Feed["id"] };
}) {
  const data = await feedModels.updateFeed(body, feedId);

  return {
    status: 200,
    data,
    message: "Feed updated",
  };
}
