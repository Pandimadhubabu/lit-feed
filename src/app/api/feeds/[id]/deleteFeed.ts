import * as feedModels from "@/models/feeds";
import { Feed } from "@/types";

export async function deleteFeed({
  params: { id },
}: {
  params: { id: Feed["id"] };
}) {
  const deletedFeed = await feedModels.deleteFeed(id);

  return {
    data: deletedFeed,
    status: 200,
    message: "Feed deleted",
  };
}
