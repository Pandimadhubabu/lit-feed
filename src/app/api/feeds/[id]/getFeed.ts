import * as feedModels from "@/models/feeds";
import { Feed } from "@/types";

export async function getFeed({
  params: { id },
}: {
  params: { id: Feed["id"] };
}) {
  const data = await feedModels.getFeed(id);

  return {
    status: 200,
    data,
    message: "Feed retrieved",
  };
}
