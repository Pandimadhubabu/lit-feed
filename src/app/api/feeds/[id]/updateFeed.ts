import * as feedModels from "@/models/feeds";
import { Feed } from "@/types";

export async function updateFeed({
  body,
  params: { id },
}: {
  body: Feed;
  params: { id: Feed["id"] };
}) {
  const updatedFeed = await feedModels.updateFeed(body, id);

  return {
    data: updatedFeed,
    status: 200,
    message: "Feed updated",
  };
}
