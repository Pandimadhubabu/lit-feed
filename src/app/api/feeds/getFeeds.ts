import * as feedModels from "@/models/feeds";

export async function getFeeds() {
  const feeds = await feedModels.getFeeds();

  return {
    message: "Feeds retrieved",
    status: 200,
    data: feeds,
  };
}
