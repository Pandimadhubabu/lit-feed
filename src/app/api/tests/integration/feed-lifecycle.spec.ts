import { testFeeds } from "@/app/api/models/testData";
import { Feed } from "@/types";
import { omit } from "../../utils";
import { getFeeds } from "../../feeds/getFeeds";
import { deleteFeed } from "../../feeds/[feedId]/deleteFeed";
import { addFeed } from "../../feeds/addFeed";
import { updateFeed } from "../../feeds/[feedId]/updateFeed";
import { getFeed } from "../../feeds/[feedId]/getFeed";

describe("feed lifecycle tests", () => {
  beforeAll(async () => {
    // Get all feeds
    const { data: allFeeds } = await getFeeds();

    // Delete all feeds
    await Promise.all(
      allFeeds.map(async (feed: Feed) => {
        const { data, message } = await deleteFeed({
          params: {
            feedId: feed.id,
          },
        });

        expect(message).toBe("Feed deleted");
        expect(data.id).toBe(feed.id);
      }),
    );
  });

  test("should be able to add a new feed", async () => {
    const { data, message } = await addFeed({
      body: testFeeds[0],
    });

    expect(message).toBe("Feed added");
    expect(omit(data, ["id"])).toEqual(omit(testFeeds[0], ["id"]));
    testFeeds[0].id = data.id;
  });

  test("should be able to update a feed", async () => {
    const nameAfterUpdate = "Updated name";

    const { data, message } = await updateFeed({
      params: {
        feedId: testFeeds[0].id,
      },
      body: {
        ...testFeeds[0],
        name: nameAfterUpdate,
      },
    });

    expect(message).toBe("Feed updated");

    expect(data.id).toBe(testFeeds[0].id);

    testFeeds[0].name = nameAfterUpdate;
  });

  test("should be able to get a feed", async () => {
    const { data, message } = await getFeed({
      params: {
        feedId: testFeeds[0].id,
      },
    });

    expect(message).toBe("Feed retrieved");
    expect(omit(data, ["id"])).toEqual(omit(testFeeds[0], ["id"]));
  });

  test("should be able to add a second feed", async () => {
    const { data, message } = await addFeed({
      body: testFeeds[1],
    });

    expect(message).toBe("Feed added");
    expect(omit(data, ["id"])).toEqual(omit(testFeeds[1], ["id"]));
    testFeeds[1].id = data.id;
  });

  test("should be able to get all feeds", async () => {
    const { data, message } = await getFeeds();

    expect(message).toBe("Feeds retrieved");

    expect(data.map((feed: Feed) => omit(feed, ["id"]))).toEqual(
      testFeeds.map((feed: Feed) => omit(feed, ["id"])),
    );
  });

  test("should be able to delete a feed", async () => {
    const { data, message } = await deleteFeed({
      params: {
        feedId: testFeeds[0].id,
      },
    });
    expect(message).toBe("Feed deleted");
    expect(data.id).toBe(testFeeds[0].id);
  });

  test("should be able to get all feeds after delete", async () => {
    const { data, message } = await getFeeds();

    expect(message).toBe("Feeds retrieved");

    expect(data.map((feed: Feed) => omit(feed, ["id"]))).toEqual([
      omit(testFeeds[1], ["id"]),
    ]);
  });

  test("should not be able to get a deleted feed", async () => {
    expect(
      getFeed({ params: { feedId: testFeeds[0].id } }),
    ).rejects.toThrowErrorMatchingSnapshot(`"Feed not found"`);
  });

  test("should not be able to update a deleted feed", async () => {
    expect(
      updateFeed({
        params: { feedId: testFeeds[0].id },
        body: { ...testFeeds[0], name: "new name" },
      }),
    ).rejects.toThrowErrorMatchingSnapshot(`"Feed not found"`);
  });

  test("should not be able to delete a deleted feed", async () => {
    expect(
      deleteFeed({ params: { feedId: testFeeds[0].id } }),
    ).rejects.toThrowErrorMatchingSnapshot(`"Feed not found"`);
  });

  test("should be able to delete a second feed", async () => {
    const { data, message } = await deleteFeed({
      params: {
        feedId: testFeeds[1].id,
      },
    });

    expect(message).toBe("Feed deleted");
    expect(data.id).toBe(testFeeds[1].id);
  });

  test("should return an empty array when getting all feeds after delete", async () => {
    const { data, message } = await getFeeds();

    expect(message).toBe("Feeds retrieved");

    expect(data).toEqual([]);
  });
});
