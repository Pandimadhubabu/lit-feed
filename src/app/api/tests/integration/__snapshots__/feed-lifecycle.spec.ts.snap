// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`feed lifecycle tests should be able to delete a second feed: "Feed not found" 1`] = `"No feed found"`;
